### PR TITLE
Add fr_dbuff_bits_in() and associated tests

### PR DESCRIPTION
**Pull request by jejones3141**
This will let us convert fr_struct_to_network() and its callers to use dbuffs

### CODE REVIEW

File: src/lib/util/dbuff.h [hunk 1]
```diff
@@ -323,8 +343,9 @@ static inline ssize_t CC_HINT(nonnull) _fr_dbuff_advance(fr_dbuff_t *dbuff, size
  */
 static inline ssize_t fr_dbuff_advance(fr_dbuff_t *dbuff, size_t inlen)
 {
-	size_t freespace = fr_dbuff_remaining(dbuff);
+	size_t freespace = fr_dbuff_remaining(dbuff);;
```
Comment (arr2036) [inline src/lib/util/dbuff.h:346]: double colon

File: src/lib/util/dbuff.h [hunk 2]
```diff
@@ -347,6 +368,7 @@ static inline ssize_t fr_dbuff_memcpy_in(fr_dbuff_t *dbuff, uint8_t const *in, s
 
 	fr_assert(!dbuff->is_const);
 
+	_fr_dbuff_bit_flush(dbuff);
```
Comment (arr2036) [inline src/lib/util/dbuff.h:371]: Wouldn't it be better to do this in the advance function?

File: src/lib/util/dbuff.h (hunk 1 above)
Comment (jejones3141) [reply src/lib/util/dbuff.h:346]: *slaps forehead* Thanks; I'll fix that.

File: src/lib/util/dbuff.h (hunk 2 above)
Comment (jejones3141) [reply src/lib/util/dbuff.h:371]: It would. I will make sure that fr_dbuff_bits_in() can work with that.

Comment (arr2036) [reply src/lib/util/dbuff.h:371]: you set remaining bits after calling advance so it should be fine

Comment (alandekok) [reply src/lib/util/dbuff.h:371]: The dictionary parsing code ensures that normal data types starts on a byte boundary.  The first `bit[]` field must also start on a byte boundary.  Then 0 or more other `bit[]` fields.  The final `bit[]` field must end on a byte boundary.

So the rest of the code can assume this is true, and blow up if it isn't true.

Comment (arr2036) [reply src/lib/util/dbuff.h:371]: I'm not sure this is the right approach in general.  Manually specifying the length is annoying.

```
fr_dbuff_bits_in(dbuff, BIT(MY_BIT_MASK_A, value), BIT(MY_BIT_MASK_B, value), BIT(MY_BIT_MASK_C, value))
```

You `|` the masks together, and take the high bit as the length of the bits you're packing. The maximum number of bits would be 64.

I think there'll be cases where we extract things like 4bit fields from structs into their own attributes, and need to re-encode them.  So being able to call fr_dbuff_bits_in and have it continue at a bit offset would still be useful.

Comment (arr2036) [reply src/lib/util/dbuff.h:371]: Seems like there's different requirements here.  I'll probably poke the function once it's pulled in, so don't worry about it for now

File: src/lib/util/dbuff.h (hunk 1 above)
Comment (jejones3141) [reply src/lib/util/dbuff.h]: Fixed.

File: src/lib/util/dbuff.h (hunk 2 above)
Comment (jejones3141) [reply src/lib/util/dbuff.h:370]: OK. Aside from the double semicolon, I've left it unchanged.

Comment (jejones3141) [reply src/lib/util/dbuff.h:370]: Pushed a change to add FR_DBUFF_BITS_IN_RETURN()

**Comment by jejones3141** [comment]
Per discussion with Alan, I'm leaving the bits_in() function for fr_struct_to_network() to use, and as part of converting fr_struct_to_network() to use dbuff, we'll convert bits_in() to use dbuff, only writing complete bytes to dbuffs, so that there's no need for fr_dbuff_bits_in().

**Pull request closed without merging**